File: 0328-odd-even-linked-list/0328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head is None or head.next is None or head.next.next is None:
            return head
        
        oddH = ListNode(-1)
        oddT = oddH
        evenH = ListNode(-1)
        evenT = evenH
        
        cur = head
        flag = False
        while cur:
            if flag == True:
                evenT.next = cur
                evenT = evenT.next
            else:
                oddT.next = cur
                oddT = oddT.next
            cur = cur.next
            flag = not flag
        
        oddT.next, evenT.next = None, None
        oddT.next = evenH.next
        
        return oddH.next
    

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        
        if(head == null || head.next == null || head.next.next == null) 
            return head;
        
        ListNode oddH = new ListNode(-1);
        ListNode oddT = oddH;
        
        ListNode evenH = new ListNode(-1);
        ListNode evenT = evenH;
        
        ListNode cur = head;
        boolean flag = false;
        
        while(cur != null){
            
            if(flag == true){
                evenT.next = cur;
                evenT = evenT.next;
            }
            else{
                oddT.next = cur;
                oddT = oddT.next;
            }
            
            cur = cur.next;
            flag = !flag;
        }
        
        oddT.next = null;
        evenT.next = null;
        
        oddT.next = evenH.next;
        return oddH.next;
        
    }
}